ci/scripts/gen-flamegraph.sh

```diff
@@ -0,0 +1,264 @@
+  echo "Should have zookeeper and kafka running"
+  ps
+  sleep 10
+  # TODO(kwannoel): `trap ERR` and upload these logs.
+  # buildkite-agent artifact upload ./zookeeper.log
+  # buildkite-agent artifact upload ./kafka.log
+}
+
+# Currently 100mil events in Kafka take up 17% of 200GB gp3 i.e. 34GB
+# 1bil records would likely exceed the current storage,
+gen_events() {
+  pushd nexmark-bench
+  nexmark-server -c
+  NEXMARK_EVENTS=$((100 * 1000 * 1000))
+  echo "Generating $NEXMARK_EVENTS events"
+  nexmark-server \
+    --event-rate 500000 \
+    --max-events "$NEXMARK_EVENTS" \
+    --num-event-generators 8 1>gen_events.log 2>&1
+  echo "Generated $NEXMARK_EVENTS events"
+  popd
+  show_kafka_topics
+}
+
+show_kafka_topics() {
+  ./kafka_2.13-3.4.0/bin/kafka-run-class.sh kafka.tools.GetOffsetShell --topic nexmark --bootstrap-server localhost:9092
+}
+
+gen_cpu_flamegraph() {
+  ./nperf flamegraph --merge-threads perf.data > perf.svg
+}
+
+############## MONITORING
+
+# TODO: promql
+monitor() {
+  sleep $((5 * 60))
+}
+
+stop_processes() {
+  # stop profiler
+  pkill nperf
+
+  # stop rw
+  pushd risingwave
+  ./risedev k
+  popd
+}
+
+############## LIB
+
+# Install artifacts + tools, configure environment
+setup() {
+  install_all
+  configure_all
+}
+
+############## MAIN
+
+main() {
+  echo "--- Machine Debug Info before Setup"
+  print_machine_debug_info
+
+  echo "--- Running setup"
+  setup
+
+  echo "--- Machine Debug Info after Setup"
+  print_machine_debug_info
+
+  echo "--- Starting kafka"
+  start_kafka
+
+  echo "--- Spawning nexmark events"
+  gen_events
+
+  echo "--- Machine Debug Info After Nexmark events generated"
+  print_machine_debug_info
+
+  echo "--- Starting up RW"
+  pushd risingwave
+  ./risedev d ci-gen-cpu-flamegraph
+  popd
+
+  echo "--- Machine Debug Info After RW Start"
+  print_machine_debug_info
+
+  echo "--- Running ddl"
+  psql -h localhost -p 4566 -d dev -U root -f risingwave/ci/scripts/sql/nexmark/ddl.sql
+
+  echo "--- Running Benchmarks"
+  # TODO(kwannoel): Allow users to configure which query they want to run.
+  psql -h localhost -p 4566 -d dev -U root -f risingwave/ci/scripts/sql/nexmark/q17.sql
+
+  echo "--- Start Profiling"
+  start_nperf
+
+  # NOTE(kwannoel): Can stub first if promql gives us issues.
+  # Most nexmark queries (q4-q20) will have a runtime of 10+ min.
+  # We can just let flamegraph profile 5min slice.
+  # TODO(kwannoel): Use promql to monitor when throughput hits 0 with 1-minute intervals.
+  echo "--- Monitoring Benchmark for 5 minutes"
+  monitor
+
+  echo "--- Benchmark finished, stopping processes"
+  stop_processes
+
+  # NOTE(kwannoel): We can only generate cpu flamegraph OR extract metrics.
+  # Running profiling will take up 30-50% of a single CPU core.
+  # So that will affect benchmark accuracy.
+  # Another solution could be to:
+  # 1. Run cpu flamegraph for first 10 mins.
+  # 2. Monitor throughput for next 10 mins.
```

Comment:
I guess there might be no need to care about the throughput seriously here. To make sure that we get an accurate result, the `nexmark-bench` pipeline from QA team is deployed on some dedicated servers and the performance can be much more stable. 🤔 If one wants to get a serious analysis of the performance changes, I presume there could be another workflow (or "GitHub PR tag" 😄) for doing this, reusing some infrastructure of the benchmark pipelines.

Reply:
> reusing some infrastructure of the benchmark pipelines.

I think we can set it up in ci first, since that's what we currently have access to. So far it seems spot instances are relatively stable, assuming that we don't run benchmarks *more frequently* than we run PRs. Running on dedicated instances may cost more :')

cc @huangjw806 for comment

Reply:
Correction from jianwei:
> The test pipeline uses demand instance, but it is of c5.2xlarge type, and the type of ci pipeline is m5.4xlarge, so there won't be much more cost.
The test pipeline is just to execute the test script, and risingwave runs in the cloud EKS, so we don't need large-type instances.

I guess it currently runs with m5.4x large instance. May not be so accurate to benchmark on it, since we usually use c5.2x large.

Generate flamegraph should still be fine.